### ISSUE TITLE
RPC Connection Handling

### ISSUE THREAD

**Issue by rgrinberg** [ocaml/dune#4355]
This ticket is born out of trying to iron the details of the connection
establishment between LSP and dune's RPC.

The issue can be described as follows. When attempts to establish a connection
to RPC, one of these must hold:

* The server is already running because the user launched dune in watch mode

* The LSP server will run the RPC server itself.

The problem with 1st option is that we must wait for the user to run the server
before providing any RPC dependant functionality. The problem with 2nd option is
that users can no longer run dune build as they would expect.

I initially proposed a solution where the editor would launch the RPC server,
and `dune build` would run separately and report errors to it. However, after
thinking about it, I noticed a very rough edge. We must have whoever is serving
RPC and the build agreeing on the environment. E.g, consider:

```sh
$ dune rpc start # starts RPC server
$ opam exec --switch=bar -- dune build -w # connects to RPC and reports errors
```

The RPC server and the build don't agree on the build environment, hence the RPC
server cannot adequately respond to any requests.

Ultimately, whoever is serving RPC requests needs to do it in the correct build
environment. Only the user knows the intended build environment, so we should
always defer to the user to run the build and serve RPC.

I propose an alternative solution that respects this fact. Essentially, we just
need to get used to the fact that editor <-> RPC connection is transient and the
user may interrupt it at any moment. I propose to alter our protocol to reflect
this:

* `$ dune rpc init` will try to reconnect if the connection is interrupted.
* The user is going to consume a stream of sessions `Session.t
  Fiber.Stream.In.t`.
* When the editor is waiting for the RPC server, a notification will appear
  telling the user to build something and serve RPC (in or out watch mode).
* When a session terminates, all pending requests terminate. The client is
  responsible to re-create subscriptions

cc @jeremiedimino @cwong-ocaml @tmattio

**Comment by rgrinberg** [comment]
Ping @jeremiedimino. If you think this approach is good, I’d rather work on this sooner rather than later as it will change some client code.

**Comment by cwong-ocaml** [comment]
I like this approach. It seems like it should be the responsibility of the client consumer to ensure that the build is running in the correct environment (so either the editor or the end-user, to be determined by the consumer).

**Comment by rgrinberg** [comment]
> so either the editor or the end-user, to be determined by the consumer

In the general case, the editor has no way of knowing unfortunately. The user could be using esy, a custom opam switch, configured his build environment through some env var tweaks. It would be nice to make this standard somehow, perhaps by adding a marker file that could hint to the editor what build environment he'd like, but the mechanism described here would still remain useful.

**Comment by cwong-ocaml** [comment]
To be clear, I think that the mechanism described here is a good idea. I guess my point of view is that the editor can be the one responsible for controlling the build environment as well (in which it should know what to do), but if it isn't, then the user should make sure that the changes are adhered to.

**Comment by jeremiedimino** [comment]
This proposal seems reasonable to me. @rgrinberg I liked your idea of having an intermediate lightweight Dune process that would answer requests that can be answered without inspecting the environment, but making that work properly probably require some refactoring of the internals of Dune as currently everything depends on `Context.t` which depends on the environment.

**Comment by rgrinberg** [comment]
>  I guess my point of view is that the editor can be the one responsible for controlling the build environment as well (in which it should know what to do),

That's quite a safe assumption to make in ecaml. Since I assume you have a standard way of launching the build.

>  I liked your idea of having an intermediate lightweight Dune process that would answer requests that can be answered without inspecting the environment

When I considered the kind requests such a lightweight process would serve, I thought of things such as `dune describe`. However, even for such simple requests, the result can be highly environment dependant, due to enabled_if for example.

So I think the simpler path to solve all these is just having the editor RPC and the build agree on the environment somehow. Nothing I propose in this ticket excludes that possibility in the future, since all I'm really doing is just making it easier to deal with transient connections. That should be useful in any situations.

**Comment by jeremiedimino** [comment]
> When I considered the kind requests such a lightweight process would serve, I thought of things such as dune describe. However, even for such simple requests, the result can be highly environment dependant, due to enabled_if for example.

Indeed. We should be able to organise the code base so the environment is never forced unless you use such features. We are going to have to do such changes if we want to take advantage of the new memo implementation for fast polling builds anyway.

> So I think the simpler path to solve all these is just having the editor RPC and the build agree on the environment somehow. Nothing I propose in this ticket excludes that possibility in the future, since all I'm really doing is just making it easier to deal with transient connections. That should be useful in any situation.

Make sense

**Comment by rgrinberg** [comment]
I'm tackling this implementation and I realize that things are a little more complicated than I originally thought. My original proposal glosses over the fact that we need `$ dune rpc init` to allow multiple sessions over a single channel (stdin/stdout).

At minimum, we need a special packet that will signal that an existing session ends. In my prototype, I chose `3:eof`, and that seems to work. Mostly because our protocol passes records rather than raw values. However, to make all of this robust,  I realized that the client should have a way to intercept all of these connection related hooks. I'd imagine a client that is able to:

* Tell the client when it is waiting for a connection (to show in the editor)
* Tell the user when an existing connection is interrupted (again, useful for UI)
* Finally, to allow the user to initialize a fresh session when it's ready.

So something like that:

```ocaml
type t =
  { on_new_session : (Initialize.t -> Handler.t -> Dune_rpc.Client.t) -> unit Fiber.t
  ; on_disconnect : (unit -> unit Fiber.t)
  }
```
